src/TorchSharp/NN/Optimizer.cs

```diff
@@ -293,563 +452,1901 @@ public static AdadeltaOptimizer Adadelta(IEnumerable<(string name, Modules.Param
+            /// <param name="momentum">Momentum factor (default: 0)</param>
+            /// <param name="dampening">Dampening for momentum (default: 0)</param>
+            /// <param name="weight_decay">Weight decay (L2 penalty) (default: 0)</param>
+            /// <param name="nesterov">Enables Nesterov momentum (default: False)</param>
+            /// <param name="maximize"></param>
+            /// <returns></returns>
+            public SGD(IEnumerable<Parameter> parameters, double lr, double momentum = 0.0, double dampening = 0, double weight_decay = 0, bool nesterov = false, bool maximize = false)
+                : this(new ParamGroup[] { new ParamGroup { Parameters = parameters } }, lr, momentum, dampening, weight_decay, nesterov, maximize)
+            {
+            }
+
+            /// <summary>
+            /// Implements stochastic gradient descent (optionally with momentum).
+            /// </summary>
+            /// <param name="parameters">Parameters to optimize. This optimizer requires the <b>named</b> parameters collection.</param>
+            /// <param name="lr">Learning rate</param>
+            /// <param name="momentum">Momentum factor (default: 0)</param>
+            /// <param name="dampening">Dampening for momentum (default: 0)</param>
+            /// <param name="weight_decay">Weight decay (L2 penalty) (default: 0)</param>
+            /// <param name="nesterov">Enables Nesterov momentum (default: False)</param>
+            /// <param name="maximize"></param>
+            /// <returns></returns>
+            public SGD(IEnumerable<ParamGroup> parameters, double lr, double momentum = 0.0, double dampening = 0, double weight_decay = 0, bool nesterov = false, bool maximize = false)
+            {
+                if (lr < 0.0) throw new ArgumentException($"Invalid learning rate: {lr}");
+                if (momentum < 0.0) throw new ArgumentException($"Invalid momentum value: {momentum}");
+                if (weight_decay < 0.0) throw new ArgumentException($"Invalid weight_decay value: {weight_decay}");
+                if (nesterov && (momentum <= 0 || dampening != 0)) throw new ArgumentException("Nesterov momentum requires a momentum and zero dampening");
+
+                var options = new Options {
+                    LearningRate = lr,
+                    InitialLearningRate = lr,
+                    dampening = dampening,
+                    maximize = maximize,
+                    momentum = momentum,
+                    nesterov = nesterov,
+                    weight_decay = weight_decay
+                };
+
+                _defaults = options;
+                _parameter_groups = new List<Modules.ParamGroup>();
+
+                foreach (var g in parameters) {
+                    add_param_group(g);
+                }
+            }
+
+            public override Tensor step(Func<Tensor> closure = null)
+            {
+                return _step<ParamGroup>(group => {
```

Comment:
I can't really check the correctness of the optimizer implementation.

A few things:

1. It needs a link to either the paper or to the PyTorch implementation
2. It needs a reproducible test for every valid options combination.

Reply:
Writing tests could be pretty easy with `TestOptimizer(Model, Optimizer, float targetLoss)` helper function that would do a reproducible run.

The problem would be to know what the target loss should be.

Reply:
>> The problem would be to know what the target loss should be.

Yes. I can think of doing it two ways:

1. Write all the unit tests in both Python and .NET and use the Python values as 'expected.' Tons of work.
2. Run the unit tests once, look at the loss and use that as the target loss.